### PR TITLE
Widget Inspector support for tracking Widget creation locations.

### PR DESCRIPTION
**Pull request by jacob314**
See
https://github.com/flutter/engine/pull/4529
for the corresponding flutter_engine cl.
and 
https://github.com/flutter/flutter/pull/13997
for the corresponding flutter_tools cl

This CL is part of implementing https://github.com/flutter/flutter-intellij/issues/1597
With this CL and a pending flutter intellij CL, widget creation locations work end to end.

### CODE REVIEW

File: packages/flutter/lib/src/widgets/widget_inspector.dart [hunk 1]
```diff
@@ -1007,3 +1021,81 @@ const TextStyle _messageStyle = const TextStyle(
...
+
+/// Interface for classes that track the source code location the their
+/// constructor was called from.
+///
+/// A [Dart Kernel Transformer](https://github.com/dart-lang/sdk/wiki/Kernel-Documentation).
+/// adds this interface to the [Widget] class when the the
+/// `--track-widget-creation` flag is passed to `flutter_tool`. Dart 2.0 is
+/// required as injecting creation locations requires a
+/// [Dart Kernel Transformer](https://github.com/dart-lang/sdk/wiki/Kernel-Documentation).
+abstract class _HasCreationLocation {
```
Comment (jacob314) [inline packages/flutter/lib/src/widgets/widget_inspector.dart]: This interface could be made public. I have erred on the side of making it private for now so no users accidentally try to implement it.

File: packages/flutter/lib/src/widgets/widget_inspector.dart [hunk 2]
```diff
@@ -1007,3 +1021,80 @@ const TextStyle _messageStyle = const TextStyle(
...
+
+/// Returns the creation location of an object if one is available.
+///
+/// Creation locations are only available for slow mode builds when
+/// the `--track-widget-creation` flag is passed to `flutter_tool`. Dart 2.0 is
+/// required as injecting creation locations requires a
+/// [Dart Kernel Transformer](https://github.com/dart-lang/sdk/wiki/Kernel-Documentation).
+///
+/// Currently creation locations are only available for [Widget] and [Element]
+_Location _getCreationLocation(Object object) {
```
Comment (jacob314) [inline packages/flutter/lib/src/widgets/widget_inspector.dart]: This is the most reasonable method to make public once we are a bit more confident in the creation locations and want to expose them to other tools.

File: packages/flutter/lib/src/widgets/widget_inspector.dart [hunk 3]
```diff
@@ -1007,3 +1021,80 @@ const TextStyle _messageStyle = const TextStyle(
...
+/// `--track-widget-creation` flag is passed to `flutter_tool`. Dart 2.0 is
+/// required as injecting creation locations requires a
+/// [Dart Kernel Transformer](https://github.com/dart-lang/sdk/wiki/Kernel-Documentation).
+abstract class _HasCreationLocation {
+  _Location get _location;
+}
+
+/// A tuple with file, line, and column number, for displaying human-readable
+/// file locations.
+class _Location {
```
Comment (jacob314) [inline packages/flutter/lib/src/widgets/widget_inspector.dart]: This location class would be made public if we choose to expose the source location API to user level Flutter code.

File: packages/flutter/lib/src/widgets/widget_inspector.dart [hunk 4]
```diff
@@ -397,6 +403,19 @@ class WidgetInspectorService {
     final Element current = selection?.currentElement;
     return _serialize(current == previousSelection?.value ? previousSelection : current?.toDiagnosticsNode(), groupName);
   }
+
+  /// Returns whether [Widget] creation locations are available.
+  ///
+  /// [Widget] creation locations are only available for slow mode builds when
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/widget_inspector.dart]: s/slow/debug/

File: packages/flutter/lib/src/widgets/widget_inspector.dart [hunk 5]
```diff
@@ -1007,3 +1026,81 @@ const TextStyle _messageStyle = const TextStyle(
...
+    this.name,
+    this.parameterLocations
+  });
+
+  /// File path of the location.
+  final String file;
+
+  /// 1-based line number.
+  final int line;
+  /// 1-based column number.
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/widget_inspector.dart]: missing blank line

**Comment by Hixie** [comment]
LGTM

Bonus points if you can figure out how to test this code. :-)

**Pull request merged**